src/ILLink.Shared/DataFlow/ValueSet.cs

```diff
@@ -9,49 +10,151 @@ namespace ILLink.Shared.DataFlow
 	public readonly struct ValueSet<TValue> : IEquatable<ValueSet<TValue>>, IEnumerable<TValue>
 		where TValue : notnull
 	{
-		public readonly HashSet<TValue>? Values;
+		// Since we're going to do lot of type checks for this class a lot, it is much more efficient
+		// if the class is sealed (as then the runtime can do a simple method table pointer comparison)
+		class SealedHashSet : HashSet<TValue>
+		{
+			public SealedHashSet () { }
+			public SealedHashSet (IEnumerable<TValue> values) : base (values) { }
+		}
+
+		public struct Enumerator : IEnumerator<TValue>, IDisposable, IEnumerator
+		{
+			readonly object? _value;
+			int _state;  // 0 before begining, 1 at item, 2 after end
+			readonly IEnumerator<TValue>? _enumerator;
+
+			internal Enumerator (object? values)
+			{
+				_state = 0;
+				if (values is SealedHashSet valuesSet) {
+					_enumerator = valuesSet.GetEnumerator ();
```

Comment:
Won't this box the enumerator?

Reply:
I didn't try to optimize the case where it falls back to the hashset. Based on measurements in illink, that case is super rare (for a simple ASP.NET app out of 8200 values inspected, only 100 has more than value in them - those would fall back to hashset)